### ISSUE TITLE
Color Border of Viewer Canvas

### ISSUE THREAD

**Issue by meeximum** [openseadragon/openseadragon#893]
Hi,

is it somehow possible to change the color of the blue border around the Viewer?

regards
Meex

**Comment by iangilman** [comment]
Where are you seeing a blue border? That sounds like the browser's focus indicator, but I'm not seeing it here even though the viewer has focus: 

http://openseadragon.github.io/examples/ui-keyboard-navigation/

**Comment by feuerste** [comment]
I can see the blue color, too. It only happens to me on Chrome after starting mouse navigation.
I attached a screenshot, which shows the 1px wide border around the black area of the image.

![capture](https://cloud.githubusercontent.com/assets/362372/19346610/8f2a9980-9144-11e6-87b5-e9e0939baf37.PNG)

**Comment by iangilman** [comment]
Okay, I see it too in Chrome. Interesting that it's in the black but not over the image... that should give a clue as to which element it's on. Anyway, I guess there is CSS one can use to hide the focus indicator?

I think we had some discussion a while ago about whether we wanted the focus indicator, since for most users it's just visual annoyance, whereas for folks who require keyboard support, it can tell you if you're in the right place. That said, this indicator is so subtle it's hardly useful, so maybe we should just get rid of it if we can.

Anyone here want to dig in to which element and what CSS needs to be applied?

**Comment by feuerste** [comment]
Actually following CSS switches off the focus indicator:

`.openseadragon-container :focus
{
    outline: none;
}`

**Comment by avandecreme** [comment]
Should we just document that instead of changing the default?

**Comment by iangilman** [comment]
@feuerste Beautiful!

@avandecreme I guess I could go either way... it'd be great if we had someone who knew a thing or two about accessibility to weigh in...

**Comment by thejohnhoffer** [comment]
@iangilman I'm not an expert: but I've recorded one video of myself tabbing through http://openseadragon.github.io/examples/ui-keyboard-navigation/ using VoiceOver and another without using VoiceOver. Without using my screen reader, I don't notice any blue focus indicator at all when tabbing through the site using only my keyboard.

![novoiceover](https://cloud.githubusercontent.com/assets/9781588/19522096/fe378b1e-95e3-11e6-9afa-07868b71c99a.gif)

![voiceover](https://cloud.githubusercontent.com/assets/9781588/19522102/01257b56-95e4-11e6-8e5e-69e66753af9c.gif)

**Comment by thejohnhoffer** [comment]
In practice, this is all that's getting my desired behavior in Chrome. 

```
.openseadragon-container:focus:not(:hover) > div { 
    border: 10px solid grey !important;
}
```

**Comment by iangilman** [comment]
I like it... seems like a good direction.

I think I'd prefer if we had an actual named class instead of using `+ div + div`. Other people and plugins add things to our divs and we don't want to accidentally style them.

**Comment by alexlande** [comment]
From a keyboard accessibility perspective, @thejohnhoffer's proposed CSS looks like a great solution.

For this case, relying on `:focus:not(:hover)` is probably good enough because the nature of the viewer is that you're hovering over it any time you're interacting with it via the mouse. For other cases where you might not hold the mouse over the element continuously, an approach similar to the modality polyfill here might make sense: https://github.com/WICG/modality.

Regarding the helper text for screen readers-- I don't think that all readers read CSS `content` (although I think VoiceOver does). If you want to make it a little more bulletproof, you might add that text via HTML (hiding it off screen until the viewer gains focus), and use `aria-describedby` to associate it with the viewer so that screen reader users will have it read to them when they select the viewer.

**Comment by iangilman** [comment]
@alexlande Awesome, thank you for weighing in on this!

@thejohnhoffer Are you up for making a patch?

**Comment by sanioka** [comment]
✅ My solution: 
`        .openseadragon-canvas:focus {
            outline: none;
        }`

**Comment by msalsbery** [comment]
> @alexlande Awesome, thank you for weighing in on this!
> 
> @thejohnhoffer Are you up for making a patch?

@iangilman 

I know this is really old, but what were you imagining being in a patch for this beyond adding class names to the control elements in the ControlDock?

**Comment by iangilman** [comment]
It seems that @thejohnhoffer had a pretty comprehensive solution proposal, and @alexlande added some thoughts, so I suppose I was thinking the patch would implement those ideas.

**Comment by msalsbery** [comment]
> It seems that @thejohnhoffer had a pretty comprehensive solution proposal, and @alexlande added some thoughts, so I suppose I was thinking the patch would implement those ideas.

It's all CSS I believe...that's why I asked - to see what code would need to be changed. I'm not sure forcing styles, especially on pseudo classes which are difficult to set from javascript, in the library is a good idea - that's something for the user in my opinion.

Adding class names to the ControlDock elements in the code would be good though...

**Comment by msalsbery** [comment]
@iangilman 

After thinking about this and researching more, I think it would be good to have at least basic accessibility stuff (as described above) built in.

It would be cool if @thejohnhoffer was around and up for a patch ;) but I wouldn't mind tackling it for my own education...

**Comment by iangilman** [comment]
I think it's good for us to control the styles inside the viewer as needed, but I agree we shouldn't be doing things that can't be overridden by the developer if they need to. For that matter, at this point we don't have a stylesheet at all... all of our CSS is done via JavaScript. Hopefully we can continue that way. That said, maybe it's worth considering adding a stylesheet if it'll make it easier for people to customize (at least in ways that make sense). I wouldn't want to do that lightly, though, as there are hundreds of apps out there using OSD not expecting to have to include a stylesheet! I suppose if it were optional... 

Anyway, great if you're up for taking a look! I haven't heard anything from @thejohnhoffer in a while... not sure what he's up to.

**Comment by msalsbery** [comment]
@iangilman from the little research I’ve done so far, it looks like we can add a stylesheet from javascript...hopefully something supported on our supported browsers

**Comment by iangilman** [comment]
Right on. Just so long as it's possible for devs to override the styles as needed!
